tests/core/test_memory.py

```diff
@@ -0,0 +1,161 @@
+
+    def forward(self, x, y):
+        out1 = self.layer1(x)
+        out2 = self.layer2(y)
+        out = self.relu(torch.cat((out1, out2), 1))
+        out = self.combine(out)
+        return out
+
+
+def test_empty_model_summary_shapes():
+    """ Test that the summary works for models that have no submodules. """
+    model = EmptyModule()
+    summary = model.summarize(mode=ModelSummary.MODE_FULL)
+    assert summary.in_sizes == []
+    assert summary.out_sizes == []
+    assert summary.param_nums == []
+
+
+@pytest.mark.parametrize(['device', 'dtype'], [
+    pytest.param(torch.device('cpu'), torch.double),
+    pytest.param(torch.device('cuda', 0), torch.float),
+    pytest.param(torch.device('cuda', 0), torch.float16),
+])
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="Test requires GPU.")
+def test_linear_model_summary_shapes(device, dtype):
+    """ Test that the model summary correctly computes the input- and output shapes. """
+    model = UnorderedModel().type(dtype).to(device)
+    model.train()
+    summary = model.summarize(mode=ModelSummary.MODE_FULL)
+    assert summary.in_sizes == [
+        [2, 10],    # layer 2
+        [2, 7],     # combine
+        [2, 3],     # layer 1
+        [2, 7],     # relu
+        UNKNOWN_SIZE,
+    ]
+    assert summary.out_sizes == [
+        [2, 2],     # layer 2
+        [2, 9],     # combine
+        [2, 5],     # layer 1
+        [2, 7],     # relu
+        UNKNOWN_SIZE,
+    ]
+    assert model.training
+    assert model.dtype == dtype
+    assert model.device == device
+
+
+def test_rnn_summary_shapes():
+    """ Test that the model summary works for RNNs. """
+    model = ParityModuleRNN()
+
+    b = 3
+    t = 5
+    i = model.rnn.input_size
+    h = model.rnn.hidden_size
+    o = model.linear_out.out_features
+
+    model.example_input_array = torch.zeros(b, t, 10)
+
+    summary = model.summarize(mode=ModelSummary.MODE_FULL)
+    assert summary.in_sizes == [
+        [b, t, i],  # rnn
+        [b, t, h],  # linear
+    ]
+    assert summary.out_sizes == [
+        [[b, t, h], [[1, b, h], [1, b, h]]],    # rnn
+        [b, t, o]                               # linear
+    ]
+
+
+def test_summary_parameter_count():
+    """ Test that the summary counts the number of parameters in every submodule. """
+    model = UnorderedModel()
+    summary = model.summarize(mode=ModelSummary.MODE_FULL)
+    assert summary.param_nums == [
+        model.layer2.weight.numel() + model.layer2.bias.numel(),
+        model.combine.weight.numel() + model.combine.bias.numel(),
+        model.layer1.weight.numel() + model.layer1.bias.numel(),
+        0,  # ReLU
+        model.unused.weight.numel() + model.unused.bias.numel(),
+    ]
+
+
+def test_summary_layer_types():
+    """ Test that the summary displays the layer names correctly. """
+    model = UnorderedModel()
+    summary = model.summarize(mode=ModelSummary.MODE_FULL)
```

Comment:
would it be cleaner to introduce `mode=ModelSummary.MODE_TYPES`

Reply:
ok, I added the parameterization in the test for all modes